### PR TITLE
Gradient-based counter-example search for properties

### PR DESCRIPTION
**Pull request by kksnowy**
This PR introduces gradient-based counter-example search for properties. It contains the following:

Compiler changes:
- A new compiler pass in the loss backend which lifts quantifiers to the highest possible levels in properties so that they can be removed easily when constructing input-differentiable loss functions in the Python backend. This compiler pass also restructures properties into boolean trees so counter-examples can be found by traversing a property's tree and searching for witnesses for leaves (via gradient descent).

Python backend changes:
- New functions which load a specification for counter-example search and prepares the required information e.g. boolean trees, input-differentiable loss functions, and quantified variable bounds for projecting perturbed points onto a valid domain in gradient descent
- A function which finds a counter-example for a property by traversing its tree and applying gradient descent on leaves.
- Updates to the ABCSampler interface and default PyTorch sampler. These now include the ability to return loss values and inputs obtained in gradient descent
- An entry function to find counter-examples for properties in a specification (only available for PyTorch at the moment)

### CODE REVIEW

**Comment by kksnowy** [comment]
As discussed last meeting, I removed the ```functionaliseResources``` step in ```compileToSearchLoss``` and removed the changes I made in ```FunctionaliseResources.hs``` (please see commit https://github.com/vehicle-lang/vehicle/pull/1167/changes/52224bbc7cbc318d376d32f5186bf6e865b34a47) but this now results in the below error when compiling to a loss function for search.

image

This error can be reproduced by running ```uv run --extra test --extra pygments python -m pytest -s tests/test_search.py```

**Comment by MatthewDaggitt** [comment]
Okay I have merged the latest changes :tada: . Nothing major - we now have to call `forceValue` before pattern matching using `toRatTensorValue`/`toBoolTensorValue` etc.

I have also fixed the `functionaliseResources` bug by handling free variables that now still exist in the `fromJSON` methods.

@kksnowy it looks like you don't have your pre-commit hooks properly configured (for code formatting etc.). Can you follow the instructions for installing the pre-commit hooks in the `CONTRIBUTING.md` file?

**Comment by kksnowy** [comment]
@MatthewDaggitt Sorry for the delay, I have done that now, and the formatting issues should now be fixed.

Also, I think there's now an issue with compiling to loss functions for search using the VehicleLoss differentiable logic. I have mentioned this in this issue https://github.com/vehicle-lang/vehicle/issues/1211 which has all of the remaining to-do's for my project.

**Comment by MatthewDaggitt** [comment]
@kksnowy at some point we should seriously think about merging this in! I guess let me know when you have tidied up the code to your satisfaction and I can give it a serious code review.

File: vehicle-python/src/vehicle_lang/_ast/_nodes.py [hunk 1]
```diff
@@ -538,3 +588,11 @@ def from_dict(cls, value: JsonValue) -> Self:
 @dataclass(frozen=True)
 class Main(Program):
     declarations: Sequence[Declaration]
+
+
+@dataclass(frozen=True)
+class SearchMain(Program):
+    """Stores boolean trees and declarations"""
```
Comment (MatthewDaggitt) [inline vehicle-python/src/vehicle_lang/_ast/_nodes.py]: If we're going to leave a comment here, could we make it a little bit more meaningful describing what relation the trees have with the program?

File: vehicle/tests/golden/errors/polarity/alternating/test.json [hunk 2]
```diff
@@ -1,5 +1,12 @@
...
-  "needs": ["alternating.vcl", "fake.onnx"]
-}
+[
+  {
+    "name": "Marabou",
+    "run": "vehicle compile queries -s alternating.vcl -f MarabouQueries --network f:fake.onnx",
+    "needs": ["alternating.vcl", "fake.onnx"]
+  },
+  {
+    "name": "DL2LossSearch",
```
Comment (MatthewDaggitt) [inline vehicle/tests/golden/errors/polarity/alternating/test.json]: I'm sorry to be a pain but could we just call the tests `Search` rather than `DL2LossSearch`. I don't think the logic used is important.

File: vehicle/src/Vehicle/Compile.hs [hunk 3]
```diff
@@ -158,19 +160,49 @@ compileToLossFunction ::
...
+  m ()
+compileToSearchLoss differentiableLogicID outputFile typedProg outputAsJSON = do
+  (booleanTrees, boolDecls, domainExtractedProg) <- convertToSearchTree typedProg
+  -- boolDecls are not converted into loss
+  lossTensorProg <- convertToLossTensors differentiableLogicID domainExtractedProg
+  let searchProg = Main (programDeclarations lossTensorProg ++ boolDecls)
+  jsonSearchProg <- convertToJSONSearchProg (booleanTrees, searchProg)
+  let outputText
+        | outputAsJSON = prettyAsJSON jsonSearchProg
+        | otherwise =
```
Comment (MatthewDaggitt) [inline vehicle/src/Vehicle/Compile.hs]: Can we put a `do` here and get rid of the `in`?

File: vehicle/src/Vehicle/Backend/LossSearch.hs [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+convertDecls ::
+  (MonadSearch m) =>
+  [Decl Builtin] ->
+  m ([BooleanTree], [Decl Builtin], [Decl Builtin])
+convertDecls = \case
+  [] -> return ([], [], [])
+  decl : decls -> do
+    (maybeBooleanTree, maybeBoolDecl, maybeDomainExtractedDecl) <- convertDecl decl
+    (booleanTrees, boolDecls, domainExtractedDecls) <- addDeclEntryToContext decl $ convertDecls decls
+    return (maybeToList maybeBooleanTree ++ booleanTrees, fromMaybe [] maybeBoolDecl ++ boolDecls, fromMaybe [] maybeDomainExtractedDecl ++ domainExtractedDecls)
```
Comment (MatthewDaggitt) [inline vehicle/src/Vehicle/Backend/LossSearch.hs]: Can we split this onto four different lines? Very long and kind of hard to read currently.

File: vehicle/src/Vehicle/Backend/LossSearch.hs [hunk 5]
```diff
@@ -0,0 +1,201 @@
+module Vehicle.Backend.LossSearch
+  ( convertToSearchTree,
+    BooleanTree (..),
+  )
+where
+
+import Control.Monad.Reader
+-- import Data.Map qualified as Map
```
Comment (MatthewDaggitt) [inline vehicle/src/Vehicle/Backend/LossSearch.hs]: Remove?

File: vehicle/src/Vehicle/Backend/LossSearch.hs [hunk 6]
```diff
@@ -0,0 +1,201 @@
+module Vehicle.Backend.LossSearch
+  ( convertToSearchTree,
+    BooleanTree (..),
```
Comment (MatthewDaggitt) [inline vehicle/src/Vehicle/Backend/LossSearch.hs]: Types usually come before functions in the export list.

File: vehicle/src/Vehicle/Backend/LossSearch.hs [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+import Vehicle.Data.Code.ForcedValue (GenericClosure (..), GenericForcedValue (..), GenericThunk (..), Thunk, emptyBoundEnv, namedBoundContextToEnv)
+import Vehicle.Data.Code.Interface
+import Vehicle.Data.MaybeTrivial
+import Vehicle.Data.Variable.Bound.Context.Name (runFreshNameBoundContextT)
+import Vehicle.Data.Variable.Bound.Context.Name.Class
+import Vehicle.Data.Variable.Bound.Context.Tensor.Instance (runFreshTensorBoundContextT)
+import Vehicle.Data.Variable.Free.Context (MonadFreeContext (..), runFreshFreeContextT)
+import Vehicle.Verify.Specification (Property, QuerySet (..))
+
+data BooleanTree
```
Comment (MatthewDaggitt) [inline vehicle/src/Vehicle/Backend/LossSearch.hs:37]: Can we add a comment on what this data type represents? The name is a bit mysterious otherwise.

File: vehicle/src/Vehicle/Backend/LossSearch.hs [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+    MonadSupply Int m
+  ) =>
+  Provenance ->
+  Identifier ->
+  DefFunctionSort ->
+  Type Builtin ->
+  QuerySet LiftedData ->
+  m (QuerySet Name, [Decl Builtin], [Decl Builtin])
+reconstructQuerySet p ident sort typ (QuerySet negated (DisjunctAll liftedData)) = do
+  reconstructedDisjuncts <- runReaderT (traverse (reconstructQueryDisjunct p ident sort typ) liftedData) (ident, p)
```
Comment (MatthewDaggitt) [inline vehicle/src/Vehicle/Backend/LossSearch.hs]: Do we actually need this reader monad if we're passing through `p` and `ident` directly? If we do, can we start to run it down lower down in the code only where it is needed?

File: vehicle/src/Vehicle/Backend/LossSearch.hs [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+  ) =>
+  LiftedData ->
+  m (Thunk Builtin, Thunk Builtin)
+reconstructExpr (quantifiers, value) = case quantifiers of
+  [] -> return (value, value)
+  (quantifier, dimsOrType, binder) : qs -> do
+    (newQuantifiedBody, newLamBody) <- addNameToContext binder $ do
+      (reconstructedQuantified, reconstructedLam) <- reconstructExpr (qs, value)
+      lv <- getBinderDepth
+      let quantifiedBody = unnormalise lv reconstructedQuantified
```
Comment (MatthewDaggitt) [inline vehicle/src/Vehicle/Backend/LossSearch.hs:169]: It feels weird that we're returning `Thunk`s from here, only to immediately renormalise them here. Can we get away with returning something of type `Expr` immediately instead? Feels like it would simplify the implementation a lot?

File: vehicle/src/Vehicle/Backend/Loss/LiftQuantifier.hs [hunk 10]
```diff
@@ -0,0 +1,300 @@
...
+  m (Property LiftedData)
+compileHardBooleanTree value = do
+  forcedValue <- forceThunk value
+  case toBoolValue forcedValue of
+    VBoolLiteral bool ->
+      return $ Trivial bool
+    VAnd (TensorOp2Args _ arg1 arg2) -> do
+      arg1' <- compileHardBooleanTree arg1
+      arg2' <- compileHardBooleanTree arg2
+      return $ andTrivial (\x y -> andBoolExpr x y) arg1' arg2'
```
Comment (MatthewDaggitt) [inline vehicle/src/Vehicle/Backend/Loss/LiftQuantifier.hs]: Can we just write `andBoolExpr` here instead of the whole lambda? Same for `or`....

File: vehicle/src/Vehicle/Backend/Loss/LiftQuantifier.hs [hunk 11]
```diff
@@ -0,0 +1,300 @@
...
+    VNot args -> do
+      errorOrResult <- runExceptT $ lowerNot noUnblocking args
+      case errorOrResult of
+        Left _ -> do
+          declProv <- ask
+          throwError $ UnableToLiftQuantifiersInProperty declProv
+        Right result -> compileHardBooleanTree result
+    VQuantifyRatTensor (quantifier, args) -> do
+      case quantifier of
+        -- If a universal quantifier is encountered, convert it to not exists not ...
```
Comment (MatthewDaggitt) [inline vehicle/src/Vehicle/Backend/Loss/LiftQuantifier.hs]: Can we put quotes around code, e.g. \`not exists not\`

File: vehicle/src/Vehicle/Backend/Loss/LiftQuantifier.hs [hunk 12]
```diff
@@ -0,0 +1,300 @@
...
+          leafDisjunction <- liftQuantifiers (Forced forcedValue, [])
+          case leafDisjunction of
+            NonTrivial disjuncts -> return $ NonTrivial $ Query $ QuerySet False disjuncts
+            Trivial bool -> return $ Trivial bool
+    VCompareIndex _ -> unblock
+    VCompareNat _ -> unblock
+    VCompareRatTensor _ -> unblock
+    VBoolIf args -> do
+      unfolded <- unfoldIf args
+      compileHardBooleanTree unfolded
```
Comment (MatthewDaggitt) [inline vehicle/src/Vehicle/Backend/Loss/LiftQuantifier.hs]: Can write this in one line as `compileHardBooleanTree =<< unfoldIf args`. Same for implies case.

File: vehicle/src/Vehicle/Backend/Loss/LiftQuantifier.hs [hunk 13]
```diff
@@ -0,0 +1,300 @@
...
+  LiftedBinders ->
+  Thunk Builtin ->
+  m (Thunk Builtin)
+updateVarLevels liftedBinders value = do
+  forcedValue <- forceThunk value
+  Forced <$> case forcedValue of
+    VFreeVar ident spine -> VFreeVar ident <$> traverseArgs (updateVarLevels liftedBinders) spine
+    VBuiltin ident spine -> VBuiltin ident <$> traverseArgs (updateVarLevels liftedBinders) spine
+    VUniverse args -> return $ VUniverse args
+    VBoundVar lv spine -> VBoundVar (lv + Lv (length liftedBinders)) <$> traverseArgs (updateVarLevels liftedBinders) spine
```
Comment (MatthewDaggitt) [inline vehicle/src/Vehicle/Backend/Loss/LiftQuantifier.hs]: Can we split this one into multiple lines?

File: vehicle-python/src/vehicle_lang/loss/_python/_translation.py [hunk 14]
```diff
@@ -12,6 +13,17 @@
 from ..._ast import _nodes as vcl
 from .._abc import ABCSampler, AnyBuiltins, Index, Tensor
 
+
+# Helper to raise a TypeError while compiling
+def invalid_type(py_ast: py.Module | py.Expression, error: TypeError) -> TypeError:
```
Comment (MatthewDaggitt) [inline vehicle-python/src/vehicle_lang/loss/_python/_translation.py]: Can we call this `raise_invalid_type`

File: vehicle-python/src/vehicle_lang/loss/_python/_translation.py [hunk 15]
```diff
@@ -36,11 +48,20 @@ class PythonTranslation(metaclass=ABCMeta):
...
+            invalid_type(py_ast, e)
+        return py_bytecode
+
+    def compile_program(
         self,
         program: vcl.Program,
         path: str | Path,
         declaration_context: dict[str, Any],
-        samplers: Mapping[str, ABCSampler[Index, Tensor]],
+        samplers: dict[str, Any],
```
Comment (MatthewDaggitt) [inline vehicle-python/src/vehicle_lang/loss/_python/_translation.py]: Did we really have to change the types here? Seems like a downgrade in typing information?

File: vehicle-python/tests/test_search.py [hunk 16]
```diff
@@ -0,0 +1,157 @@
...
+    return torch_module, loss_module
+
+
+def test_pytorch_search_bounded() -> None:
+    torch, loss_pt = require_pytorch()
+
+    spec_path = Path(__file__).parent / "data" / "test_bounded.vcl"
+    declarations = ["network", "bounded"]
+
+    # A network that calculates y = 2x + bias
```
Comment (MatthewDaggitt) [inline vehicle-python/tests/test_search.py:32]: In this and the test below, what is `bias`? I can't see a variable here, is it 0?

File: vehicle-python/src/vehicle_lang/loss/_search_tree.py [hunk 17]
```diff
@@ -0,0 +1,176 @@
...
+    Returns:
+        A list of samples representing a counter-example or witness.
+        The list is empty if a counter-example/witness has not been found.
+    """
+
+    def traverse_boolean_expr(
+        node: BooleanExpression,
+    ) -> tuple[bool, Sequence[dict[str, Any]]]:
+
+        if isinstance(node, NonTrivialQuery):
```
Comment (MatthewDaggitt) [inline vehicle-python/src/vehicle_lang/loss/_search_tree.py:42]: This method is very big. Could we split it up into `traverse_conjunct`, `traverse_disjunct` etc?

File: vehicle-python/src/vehicle_lang/loss/_common.py [hunk 18]
```diff
@@ -2,33 +2,133 @@
...
-def load_loss_specification(
+def load_loss_ast(
+    path: str | Path,
+    mode: LossMode,
+    logic: DifferentiableLogic,
+    declarations: Iterable[DeclarationName],
+) -> Program:
+    """Load a loss AST for training or search."""
+
+    program = load_ast(
```
Comment (MatthewDaggitt) [inline vehicle-python/src/vehicle_lang/loss/_common.py]: This can just be `return load_ast(...)`

File: vehicle-python/src/vehicle_lang/loss/_common.py [hunk 19]
```diff
@@ -2,33 +2,133 @@
...
+    translator: PythonTranslation,
+) -> dict[str, Any]:
+    """Translate a loss function using a provided backend factory."""
+
+    compiled = translator.compile_program(
+        program=program,
+        path=path,
+        declaration_context=declaration_context,
+        samplers=samplers,
+    )
```
Comment (MatthewDaggitt) [inline vehicle-python/src/vehicle_lang/loss/_common.py:80]: Same here

**Review by MatthewDaggitt — CHANGES_REQUESTED** [review]
A few minor comments, but nothing too major hopefully 🤞

**Comment by MatthewDaggitt** [comment]
Ooh and can we add a CHANGELOG entry describing the addition of the new search procedure to the API?

File: vehicle-python/tests/test_search.py (hunk 16 above)
Comment (kksnowy) [reply vehicle-python/tests/test_search.py:32]: Looking at the documentation for the PyTorch `Linear` layer, I think a random additive bias is applied by default? I have explicitly set `bias=True` when creating the layer to make this clearer.

**Review by MatthewDaggitt — APPROVED** [review]
